src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1783,15 +1832,65 @@ RemoveCoordinatorPlacement(List *placementList)
 
 
 /*
- * SingleShardSelectTaskList generates a task for single shard select query
+ * SingleShardTaskList generates a task for single shard query
  * and returns it as a list.
  */
 static List *
-SingleShardSelectTaskList(Query *query, uint64 jobId, List *relationShardList,
-						  List *placementList, uint64 shardId,
-						  bool parametersInQueryResolved)
+SingleShardTaskList(Query *query, uint64 jobId, List *relationShardList,
+					List *placementList, uint64 shardId,
+					bool parametersInQueryResolved)
 {
-	Task *task = CreateTask(READ_TASK);
+	TaskType taskType = READ_TASK;
+	char replicationModel = 0;
+
+	if (query->commandType != CMD_SELECT)
+	{
+		List *rangeTableList = NIL;
+		ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+
+		RangeTblEntry *updateOrDeleteRTE = ExtractResultRelationRTE(query);
+		Assert(updateOrDeleteRTE != NULL);
+
+		CitusTableCacheEntry *modificationTableCacheEntry = GetCitusTableCacheEntry(
+			updateOrDeleteRTE->relid);
+		char modificationPartitionMethod = modificationTableCacheEntry->partitionMethod;
+
+		if (modificationPartitionMethod == DISTRIBUTE_BY_NONE &&
+			SelectsFromDistributedTable(rangeTableList, query))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot perform select on a distributed table "
+								   "and modify a reference table")));
+		}
+
+		taskType = MODIFY_TASK;
+		replicationModel = modificationTableCacheEntry->replicationModel;
+	}
+
+	if (taskType == READ_TASK && query->hasModifyingCTE)
+	{
+		/* assume ErrorIfQueryHasUnroutableModifyingCTE checked query already */
+
+		CommonTableExpr *cte = NULL;
+		foreach_ptr(cte, query->cteList)
+		{
+			Query *cteQuery = (Query *) cte->ctequery;
+
+			if (cteQuery->commandType != CMD_SELECT)
+			{
+				RangeTblEntry *updateOrDeleteRTE = ExtractResultRelationRTE(cteQuery);
+				CitusTableCacheEntry *modificationTableCacheEntry =
+					GetCitusTableCacheEntry(
+						updateOrDeleteRTE->relid);
```

Comment:
I'm wondering whether all this code can be moved to a more advanced version of ExtractResultRelationRTE which descends into modifying CTEs, and returns InvalidOid if there is not result relation OID.

Reply:
bump

Reply:
It gets weird because there can be multiple outputs